src/azure/azureResourceController.ts

```diff
@@ -0,0 +1,68 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Location } from '@azure/arm-subscriptions';
+import { ResourceGroup } from '@azure/arm-resources';
+import { Server } from '@azure/arm-sql';
+import * as mssql from 'vscode-mssql';
+import * as azureUtils from './utils';
+
+export class AzureResourceController {
+
+	constructor(
+		private _subscriptionClientFactory: azureUtils.SubscriptionClientFactory = azureUtils.defaultSubscriptionClientFactory,
+		private _resourceManagementClientFactory: azureUtils.ResourceManagementClientFactory = azureUtils.defaultResourceManagementClientFactory,
+		private _sqlManagementClientFactory: azureUtils.SqlManagementClientFactory = azureUtils.defaultSqlManagementClientFactory) {
+	}
+
+	/**
+	 * Returns Azure locations for given session
+	 */
+	public async getLocations(session: mssql.IAzureAccountSession): Promise<Location[]> {
+		const subClient = this._subscriptionClientFactory(session.token);
+		if (session.subscription?.subscriptionId) {
+			const locationsPages = await subClient.subscriptions.listLocations(session.subscription.subscriptionId);
+			let locations = await azureUtils.getAllValues(locationsPages, (v) => v);
+			return locations.sort((a, b) => (a.name || '').localeCompare(b.name || ''));
+		} else {
+			throw new Error('Invalid session');
+		}
+	}
+
+	/**
+	 * Creates or updates a Azure SQL server for given subscription, resource group and location
+	 */
+	public async createOrUpdateServer(
+		subscriptionId: string,
+		resourceGroupName: string,
+		serverName: string,
+		parameters: Server,
+		token: mssql.Token): Promise<string | undefined> {
+		if (subscriptionId && resourceGroupName) {
+			const sqlClient = this._sqlManagementClientFactory(token, subscriptionId);
+			if (sqlClient) {
+				const result = await sqlClient.servers.beginCreateOrUpdateAndWait(resourceGroupName,
+					serverName, parameters);
+
+				return result.fullyQualifiedDomainName;
```

Comment:
what if the result is undefined?

Reply:
I don't think that happens. Azure APIs throw error if fails to create server